core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1151,13 +1151,22 @@ public static boolean equalSansNullability(
       return true;
     }
 
-    if (type1.isNullable() == type2.isNullable()) {
+    if (isAtomic(type1) && isAtomic(type2) && (type1.isNullable() == type2.isNullable())) {
       // If types have the same nullability and they weren't equal above,
       // they must be different.
       return false;
     }
-    return type1.equals(
-        factory.createTypeWithNullability(type2, type1.isNullable()));
+    if (isArray(type1) && isArray(type2)) {
+      return equalSansNullability(factory, type1.getComponentType(), type2.getComponentType());
+    } else if (isMap(type1) && isMap(type2)) {
+      MapSqlType mType1 = (MapSqlType) type1;
```

Comment:
I would suggest to add two new methods `equalAsArraySansNullability` and `equalsAsMapSansNullability`, because for most of the cases when we invoke these methods, we already know if the type to compare is a map/array/struct, add new methods simplify the original impl(which is the most common case).

Reply:
BTW, can you share why you need to compare map and array types ? Seems the case is rare from my side.

Reply:
> I would suggest to add two new methods `equalAsArraySansNullability` and `equalsAsMapSansNullability`, because for most of the cases when we invoke these methods, we already know if the type to compare is a map/array/struct, add new methods simplify the original impl(which is the most common case).

Thanks for your advice, I'll change this latter.

Reply:
> BTW, can you share why you need to compare map and array types ? Seems the case is rare from my side.

A case like this: array(12L) will be recognized as "INTEGER ARRAY, NOT NULL", but the col i_ids is "BIGINT ARRAY NULLABLE", so calcite will add a cast. But the source/targe nullability is still different, and we don't want to consider nullability.

```
INSERT TABLE new_t
SELECT id,
       array(12) AS i_ids,
FROM t
```

Reply:
Reasonable, in `TypeCoercionImpl#querySourceCoercion` we may need to take the array element nullability into account.